File: lstm/zig/lstm.zig
const LSTMCell = struct {
  units: usize,
  kernel: [4][usize],
  bias: [4]f32,
};

fn apply(cell: *LSTMCell, inputs: []const f32, h_t_prev: []const f32, c_t_prev: []const f32, h_t: []f32, c_t: []f32) void {
  // Apply the input gate
  i_t := sigmoid(inputs.dot(cell.kernel[0]) + cell.bias[0])

  // Apply the forget gate
  f_t := sigmoid(inputs.dot(cell.kernel[1]) + cell.bias[1])

  // Apply the cell gate
  c_t_hat := tanh(inputs.dot(cell.kernel[2]) + cell.bias[2])

  // Apply the output gate
  o_t := sigmoid(inputs.dot(cell.kernel[3]) + cell.bias[3])

  // Update the cell state
  c_t = f_t * c_t_prev + i_t * c_t_hat

  // Output the hidden state
  h_t = o_t * tanh(c_t)
}

pub fn main() void {
  // Initialize the LSTM cell
  const cell = LSTMCell(10)

  // Create the input and output sequences
  const inputs = [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
  const outputs = [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]

  // Initialize the hidden state and cell state
  var h_t_prev: []f32 = []f32(1)
  var c_t_prev: []f32 = []f32(1)

  // Apply the LSTM cell for each time step
  for t in 0..10 {
    // Apply the LSTM cell
    cell.apply(&inputs[t..], &h_t_prev, &c_t_prev, &h_t, &c_t)

    // Update the hidden state and cell state
    h_t_prev = h_t
    c_t_prev = c_t
  }

  // Print the output sequence
  for i in 0..len(h_t) {
    print(h_t[i])
  }
}

File: lstm/rust/lstm.rs
struct LSTMCell {
    units: usize,
    kernel: [[f32; 10]; 4],
    bias: [f32; 4],
  }
  
  impl LSTMCell {
    pub fn apply(&self, inputs: &[f32], h_t_prev: &[f32], c_t_prev: &[f32], h_t: &mut [f32], c_t: &mut [f32]) {
      // Apply the input gate
      let i_t = sigmoid(inputs.dot(&self.kernel[0]) + &self.bias[0]);
  
      // Apply the forget gate
      let f_t = sigmoid(inputs.dot(&self.kernel[1]) + &self.bias[1]);
  
      // Apply the cell gate
      let c_t_hat = tanh(inputs.dot(&self.kernel[2]) + &self.bias[2]);
  
      // Apply the output gate
      let o_t = sigmoid(inputs.dot(&self.kernel[3]) + &self.bias[3]);
  
      // Update the cell state
      c_t.copy_from_slice(&f_t * &c_t_prev + &i_t * &c_t_hat);
  
      // Output the hidden state
      h_t.copy_from_slice(&o_t * &tanh(&c_t));
    }
  }
  
  fn main() {
    // Initialize the LSTM cell
    let cell = LSTMCell::new(10);
  
    // Create the input and output sequences
    let inputs = vec![0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0];
    let outputs = vec![0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0];
  
    // Initialize the hidden state and cell state
    let mut h_t_prev = vec![0.0];
    let mut c_t_prev = vec![0.0];
  
    // Apply the LSTM cell for each time step
    for t in 0..10 {
      // Apply the LSTM cell
      cell.apply(&inputs[t ..], &h_t_prev, &c_t_prev, &mut h_t, &mut c_t);
  
      // Update the hidden state and cell state
      h_t_prev.copy_from_slice(&h_t);
      c_t_prev.copy_from_slice(&c_t);
    }
  
    // Print the output sequence
    println!("{:?}", h_t);
  }
  